lib/loader/ast/section.cpp

```diff
@@ -217,26 +218,38 @@ Expect<void> Loader::loadSection(AST::TagSection &Sec) {
 }
 
 Expect<void> Loader::loadSection(AST::Component::ComponentSection &Sec) {
-  auto ResPreamble = Loader::loadPreamble();
-  if (!ResPreamble) {
-    return Unexpect(ResPreamble);
-  }
-  auto WasmMagic = ResPreamble->first;
-  auto Ver = ResPreamble->second;
-  if (unlikely(Ver != ComponentVersion)) {
-    return logLoadError(ErrCode::Value::MalformedVersion, FMgr.getLastOffset(),
-                        ASTNodeAttr::Component);
-  }
-  auto NestedComp = std::make_shared<AST::Component::Component>();
-  NestedComp->getMagic() = WasmMagic;
-  NestedComp->getVersion() = {Ver[0], Ver[1]};
-  NestedComp->getLayer() = {Ver[2], Ver[3]};
-  if (auto Res = loadComponent(*NestedComp); !Res) {
-    spdlog::error(ErrInfo::InfoAST(ASTNodeAttr::Component));
-    return Unexpect(Res);
-  }
-  Sec.getContent() = NestedComp;
-  return {};
+  return loadSectionContent(Sec, [this, &Sec]() -> Expect<void> {
+    auto ExpectedSize = Sec.getContentSize();
+    auto StartOffset = FMgr.getOffset();
+
+    auto ResPreamble = Loader::loadPreamble();
+    if (!ResPreamble) {
+      spdlog::error(ErrInfo::InfoAST(ASTNodeAttr::Component));
+      return Unexpect(ResPreamble);
+    }
+    auto WasmMagic = ResPreamble->first;
+    auto Ver = ResPreamble->second;
+    if (unlikely(Ver != ComponentVersion)) {
+      return logLoadError(ErrCode::Value::MalformedVersion,
+                          FMgr.getLastOffset(), ASTNodeAttr::Component);
+    }
+    auto NestedComp = std::make_shared<AST::Component::Component>();
+    NestedComp->getMagic() = WasmMagic;
+    NestedComp->getVersion() = {Ver[0], Ver[1]};
+    NestedComp->getLayer() = {Ver[2], Ver[3]};
+
+    auto Offset = FMgr.getOffset();
+
+    if (auto Res =
+            loadComponent(*NestedComp, ExpectedSize - (Offset - StartOffset));
```

Comment:
I think there may be a situation that `ExpectedSize` < `Offset` - `StartOffset`?
Consider a situation that the section size was loaded and set as size `10` in `Loader::loadSectionContent<T>()`, and `StartOffset` is `50` when entering this function, and the `Offset` is `70` after above loading. Then the evaluate of `ExpectedSize - (Offset - StartOffset)` will underflow in a `uint32_t`.

Reply:
Indeed, I push a fix for this, there are two place use the same idea. (nested module and component)